homework/ba-nguyen/lesson-7/javascript-2.md

```diff
@@ -0,0 +1,188 @@
+1. **Write a curry function that can do the following:**
+
+   // Modify the below function
+
+   ```bash
+    function curry(f: any) {
+        const numbs: number[] = []
+        const helper = (arg: number) => {
+            if(numbs.length === f.length - 1) {
+                return f(...numbs, arg)
+            } else {
+                numbs.push(arg)
+            }
+            return helper
+        }
+
+        return helper
+    }
+   ```
+
+   // Do not modify the below function
+
+   ```bash
+   function main() {
+       const sum = (a, b) => {
+           return a + b;
+       };
+       const curriedSum = curry(sum);
+       curriedSum(1)(2); // returns 3
+   }
+
+   ```
+
+2. **Write a function that can memoize the result of a function:**
+
+   // Modify the below function
+
+   ```bash
+   type ObjectType = { [key: string]: number };
+   function memoize(func: (a: number, b: number) => number) {
```

Comment:
LGTM but you can make it cover more case. Example:

```js
function memoize(func) {
    const cache = {};
    return function (...agrs) {
      if(cache[agrs.join('')]) return cache[agrs.join('')]
      return cache[agrs.join('')] = func.apply(this, [...agrs])
    };
}
```

Reply:
for this question i memorize by adjacency and when i solve any algorithm i always use naive code, i don't want to use methods build in js

Reply:
no problem, I understand your point. But you should code with js, don't use data type to solve js assignment

Reply:
yeah, in interview, we should use pure js

Reply:
btw, this is author answer

```js
function memoize(func) {
    const cache = {};
    return function(...args) {
        if (cache[args]) {
            return cache[args];
        }
        const result = func(...args);
        cache[args] = result;
        return result;
    }
}
```